Add ENABLE macro for WebProcess cache

pvollan opened the pull request:
#### e42376bfd40b7595d9e3c77035e4a2e4e71fd10b
<pre>
Add ENABLE macro for WebProcess cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=261952">https://bugs.webkit.org/show_bug.cgi?id=261952</a>
rdar://115896548

Reviewed by Brent Fulgham.

In <a href="https://commits.webkit.org/266267@main">https://commits.webkit.org/266267@main</a>, the WebProcess cache was enabled on iOS,
but also enabled on watchOS and tvOS, which was not intentional. Fix this by adding
a ENABLE macro for the feature.

* Source/WTF/wtf/PlatformEnable.h:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WTF/wtf/PlatformEnableWinCairo.h:
* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::updateCapacity):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::prepareToSuspend):

Canonical link: <a href="https://commits.webkit.org/268334@main">https://commits.webkit.org/268334@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9636b67d9a4713a8240e29a26b0fa81126fda826

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19339 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19757 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20356 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21228 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18095 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23026 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19882 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19727 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19556 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19605 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16817 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22097 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16791 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17605 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23932 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/16788 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17853 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17780 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21899 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/18674 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18369 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15555 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/22731 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17503 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/5625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4632 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21863 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/23982 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18199 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5361 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WTF/wtf/PlatformEnable.h:
```diff
@@ -1030,6 +1030,14 @@
+    || PLATFORM(GTK) \
+    || PLATFORM(WPE) \
+    || PLATFORM(WIN))
+#define ENABLE_WEBPROCESS_CACHE 1
```
This is okay, but I've learned that the intended behavior would be to have the MAC and IOS bits set in `PlatformEnableCocoa.h`, the WIN bit in `PlatformEnableWinCairo.h` and this file take the form:

```
#if !defined(ENABLE_WEBPROCESS_CACHE)
#define ENABLE_WEBPROCESS_CACHE 0
#endif
```

This doesn't work perfectly since GTK/WPE does not have a dedicated PlatformEnablement file at present, which messes up this format.

I suggest you do the following:

1. Put the MAC/IOS enablement in `PlatformEnableCocoa.h`
2. Put the WIN enablement in `PlatformEnableWinCairo.h`
3. Have this section become this:

```
#if !defined(ENABLE_WEBPROCESS_CACHE) && (PLATFORM(GTK) || PLATFORM(WPE))
#define ENABLE_WEBPROCESS_CACHE 1
#endif

#if !defined(ENABLE_WEBPROCESS_CACHE)
#define ENABLE_WEBPROCESS_CACHE 0
#endif
```

Then we (rather, *I* or Igalia people) can clean up the GTK stuff in the future

pvollan replied:
Ah, that is a good point! Fixed in latest patch.

Thanks for reviewing!

brentfulgham reviewed (changes requested):
I think this patch is fine as-is, but I'd like to ask you to take one further step and move the enablement into the appropriate platform enablement header files. This will make future cleanups easier.

brentfulgham reviewed (approved):
Thank you for adjusting the flags. I know it was annoying! r=me

pvollan committed:
Add ENABLE macro for WebProcess cache
https://bugs.webkit.org/show_bug.cgi?id=261952
rdar://115896548

Reviewed by Brent Fulgham.

In https://commits.webkit.org/266267@main, the WebProcess cache was enabled on iOS,
but also enabled on watchOS and tvOS, which was not intentional. Fix this by adding
a ENABLE macro for the feature.

* Source/WTF/wtf/PlatformEnable.h:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WTF/wtf/PlatformEnableWinCairo.h:
* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::updateCapacity):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::prepareToSuspend):

Canonical link: https://commits.webkit.org/268334@main

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1609,6 +1609,11 @@ void WebProcess::prepareToSuspend(bool isSuspensionImminent, MonotonicTime estim
 #endif
 
+#if !ENABLE(WEBPROCESS_CACHE)
+    if (!m_suppressMemoryPressureHandler)
```
It is confusing to add new code and apparently change behavior in a patch that is meant to simply introduce a new build flag.

pvollan replied:
That's a good point, Chris! I had to re-introduce this code, since it was removed by https://commits.webkit.org/266267@main.

Thanks for reviewing!
